### PR TITLE
Send unhandledrejection events to the main thread

### PR DESCRIPTION
**Pull request by tlively**
We send uncaught exceptions to the main thread where they can be handled in a
central location, but we did not previously send unhandledrejection events that
are created when a promise is rejected with no handler. Add a handler in
worker.js that intercepts unhandledrejection events and throws them as errors so
they are propagated to the main thread in the usual manner.

The way this works varies slightly between Node and the browser, so add tests
for both cases.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #17535** <a href="https://app.graphite.dev/github/pr/emscripten-core/emscripten/17535" target="_blank">Graphite</a>  👈
    * **PR #17536** <a href="https://app.graphite.dev/github/pr/emscripten-core/emscripten/17536" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/emscripten-core/emscripten/17535?utm_source=stack-comment).

File: src/worker.js [hunk 1]
```diff
@@ -105,6 +105,14 @@ Module['instantiateWasm'] = (info, receiveInstance) => {
 }
 #endif
 
+// Turn unhandled rejected promises into errors so that the main thread will be
+// notified about them.
+self.onunhandledrejection = (e) => {
+  if (typeof(e.reason) !== 'undefined') {
```
Comment (kripken) [inline src/worker.js:111]: When is the reason undefined? (Wouldn't we still want to throw an error then?)

File: src/worker.js [hunk 2]
```diff
@@ -105,6 +105,14 @@ Module['instantiateWasm'] = (info, receiveInstance) => {
 }
 #endif
 
+// Turn unhandled rejected promises into errors so that the main thread will be
+// notified about them.
+self.onunhandledrejection = (e) => {
```
Comment (kripken) [inline src/worker.js:110]: Might be better to use `addEventListener` here so it doesn't interfere with any other global handler the user may have added. We don't do that for `onmessage` etc., which is risky as well, but we already assume we handle all the messaging so maybe that's ok. But users may have their own error handling for promises.

File: test/common.py [hunk 3]
```diff
@@ -1532,7 +1532,7 @@ def run_browser(self, html_file, expected=None, message=None, timeout=None, extr
             if extra_tries > 0:
               print('[test error (see below), automatically retrying]')
               print(e)
-              return self.run_browser(html_file, message, expected, timeout, extra_tries - 1)
+              return self.run_browser(html_file, expected, message, timeout, extra_tries - 1)
```
Comment (kripken) [inline test/common.py:1535]: Is this related to this PR?

File: src/worker.js (hunk 2 above)
Comment (sbc100) [reply src/worker.js:110]: I don't think we provide any way for the user to add event listeners in the workers.js do we?    worker.js is kind of designed to "own" then entire works, not run alongside random user code I think?

Comment (sbc100) [reply src/worker.js:110]: I don't think we provide any way for the user to add event listeners in the workers.js do we?    worker.js is kind of designed to "own" then entire works, not run alongside random user code I think?

File: test/common.py (hunk 3 above)
Comment (tlively) [reply test/common.py:1535]: No, just a drive-by fix for an issue I noticed while running browser tests.

File: test/pthread/test_pthread_unhandledrejection.post.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  if (ENVIRONMENT_IS_NODE) {
+    process.on('error', (e) => {
+      if (e === 'rejected!') {
+        console.log('passed');
+      }
+    });
+  } else {
+    addEventListener('error', (e) => {
+      if (e.message === 'Uncaught rejected!') {
+        console.log('passed');
```
[the remark this answers is not in the document]
Comment (tlively) [reply test/pthread/test_pthread_unhandledrejection.post.js:11]: Yeah, this logging doesn't do anything as far as the test harness is concerned, but I think it's useful for trying to run the test manually.

File: src/worker.js (hunk 1 above)
Comment (tlively) [reply src/worker.js:111]: Yeah, I think you're right that maybe we should just propagate the error unconditionally and unchanged. @sbc100, did you have ideas about what would be best to do here?

Comment (sbc100) [reply src/worker.js:111]: Have you ever seen the reason be undefined?   If not, perhaps its reasonable to assert that there is a reason and then throw it?

Comment (tlively) [reply src/worker.js:111]: Yes, it seems that if you do `Promise.reject("foo")`, then there is no reason in the event object. From what I've seen, all the system-generated promise rejections have reasons, though.

Comment (sbc100) [reply src/worker.js:111]: Right... but when if happens we wouldn't want to ignore it would we?  Don't we still want to propagate some kind of error?

Comment (tlively) [reply src/worker.js:111]: maybe if there is a `reason` field we should propagate that and otherwise we should propagate the event object directly?

Comment (kripken) [reply src/worker.js:111]: That sounds good. Or if there is no reason we could just propagate "rejected" or something generic. The event might be better as it might contain some info I guess.

File: src/worker.js (hunk 2 above)
Comment (kripken) [reply src/worker.js:110]: `--pre-js` code ends up evalled into the worker.js scope, I think? So any hooks there would end up relevant.

Comment (tlively) [reply src/worker.js:110]: I guess the user is more likely to want to add a promise rejection handler than an onmessage handler? But if they do, they probably want to override this one, so maybe they can do that in --post-js and it will be ok?

Comment (kripken) [reply src/worker.js:110]: Hmm, good point, maybe the user being able to override this one is better actually. sgtm.

Comment (kripken) [reply src/worker.js:110]: This is a general change to all pthreads-using applications, though, which is what worries me. Gating this on WasmFS would be safer, but odd. Perhaps we should at least document this in the changelog.

**Comment by tlively** [comment]
The other.test_run_wasi_sdk_output failure doesn't look related:

```
wasm-ld: error: cannot open /root/emsdk/upstream/lib/clang/16.0.0/lib/wasi/libclang_rt.builtins-wasm32.a: No such file or directory
```

File: src/worker.js (hunk 2 above)
Comment (sbc100) [reply src/worker.js:110]: Surely this is good change that all programs will want.   Without this, such errors go un-detected and un-reported.  I see no reason to make this wasmfs related given that it relates to all JS promises.

Comment (kripken) [reply src/worker.js:110]: Agreed. Changelog mention seems enough.

**Comment by tlively** [comment]
This should be ready for another look. I now throw the event reason if there is one, and otherwise throw the event object itself so that errors are always reported somehow.

File: src/worker.js [hunk 5]
```diff
@@ -105,6 +105,15 @@ Module['instantiateWasm'] = (info, receiveInstance) => {
...
+  if (typeof e.reason !== 'undefined') {
+    throw e.reason;
+  }
+  throw e;
```
Comment (sbc100) [inline src/worker.js:114]: This can just be `throw e.reason ? e.reason : e`?

Comment (sbc100) [reply src/worker.js:114]: Actually you do this these days I think: `throw e.reason ?? e`

Comment (tlively) [reply src/worker.js]: done.

**Comment by tlively** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/emscripten-core/emscripten/17535?utm_source=stack-comment).


Job Reference: AxYp7AbVmnyKzGFPh9SB
